### PR TITLE
hostnamed: display firmware version

### PR DESCRIPTION
**Pull request by cerebro1**
Updated hostnamectl to print firmware version
updated hostnamed to fetch bios_version from /sys/class/dmi/id/bios_version

This is a continued work from https://github.com/systemd/systemd/pull/23068
issue : https://github.com/systemd/systemd/issues/22728 

@bluca @anitazha @poettering 

Output on my system :
```
sonalis@cere:~/outreachyspring2022/systemd [hostname_outreachy] $ cat build/test/sys/class/dmi/id/bios_version 
7IET31WW (1.12 )
```
Output of hostnamectl : 
```
sonalis@cere:~/outreachyspring2022/systemd [hostname_outreachy] $ ./build/hostnamectl
 Static hostname: cere
       Icon name: computer-laptop
         Chassis: laptop 💻
      Machine ID: e7250c9873e54839aa321e08d41dbf30
         Boot ID: 5029ec53bf7b45caa8a2ff2490887a70
Operating System: Ubuntu 18.04.3 LTS
          Kernel: Linux 4.15.0-175-generic
    Architecture: x86-64
```
Still value of Firmware Version not displayed by hostnamectl. Property return empty.

### CODE REVIEW

File: src/hostname/hostnamed.c [hunk 1]
```diff
@@ -208,21 +208,20 @@ static int get_hardware_model(char **ret) {
...
         r = sd_device_new_from_syspath(&device, "/sys/class/dmi/id");
         if (r < 0)
                 return log_debug_errno(r, "Failed to open /sys/class/dmi/id device, ignoring: %m");
 
-        (void) sd_device_get_sysattr_value(device, "product_serial", &s);
-        if (isempty(s))
-                /* Fallback to board serial */
-                (void) sd_device_get_sysattr_value(device, "board_serial", &s);
-
+        if (dmi_data)
```
Comment (bluca) [inline src/hostname/hostnamed.c]: no need for this `if`, there's an assert above

File: src/hostname/hostnamed.c [hunk 2]
```diff
@@ -235,6 +234,19 @@ static int get_hardware_serial(char **ret) {
         return !isempty(s);
 }
 
+static int get_hardware_serial(char **ret) {
+         int r;
+         r = get_hardware_firmware_data("product_serial", ret);
```
Comment (bluca) [inline src/hostname/hostnamed.c:239]: style nit: empty line after variables declaration block

File: src/hostname/hostnamed.c [hunk 3]
```diff
@@ -235,6 +234,19 @@ static int get_hardware_serial(char **ret) {
...
+         if (r < 0) {
```
Comment (bluca) [inline src/hostname/hostnamed.c]: no braces around single-line if blocks

File: src/hostname/hostnamed.c [hunk 4]
```diff
@@ -208,21 +208,20 @@ static int get_hardware_model(char **ret) {
...
         if (r < 0)
                 return log_debug_errno(r, "Failed to open /sys/class/dmi/id device, ignoring: %m");
 
-        (void) sd_device_get_sysattr_value(device, "product_serial", &s);
-        if (isempty(s))
-                /* Fallback to board serial */
-                (void) sd_device_get_sysattr_value(device, "board_serial", &s);
-
+        if (dmi_data)
+                (void) sd_device_get_sysattr_value(device, dmi_data, &s)
```
Comment (bluca) [inline src/hostname/hostnamed.c]: missing `;` at the end of the line

**Comment by bluca** [comment]
> Output on my system : `sonalis@cere:~/outreachyspring2022/systemd [hostname_outreachy] $ cat build/test/sys/class/dmi/id/bios_version 7IET31WW (1.12 )` Output of hostnamectl : `sonalis@cere:~/outreachyspring2022/systemd [hostname_outreachy] $ ./build/hostnamectl Static hostname: cere Icon name: computer-laptop Chassis: laptop 💻 Machine ID: e7250c9873e54839aa321e08d41dbf30 Boot ID: 5029ec53bf7b45caa8a2ff2490887a70 Operating System: Ubuntu 18.04.3 LTS Kernel: Linux 4.15.0-175-generic Architecture: x86-64` Still value of Firmware Version not displayed by hostnamectl. Property return empty.

Seems to work fine here on my machine (in nspawn):

```
root@image:~# cat /sys/class/dmi/id/bios_version
N27ET35W (1.21 )
root@image:~# busctl get-property org.freedesktop.hostname1 /org/freedesktop/hostname1 org.freedesktop.hostname1 FirmwareVersion
s "N27ET35W (1.21 )"
root@image:~# hostnamectl
   Static hostname: n/a                             
Transient hostname: image
         Icon name: computer
        Machine ID: b198cd1a7c544cd09843480cfa055357
           Boot ID: 96aaafdfc98f49f0ba3ade1957bfd2aa
    Virtualization: systemd-nspawn
  Operating System: Debian GNU/Linux bookworm/sid   
            Kernel: Linux 5.10.0-13-amd64
      Architecture: x86-64
  Firmware Version: N27ET35W (1.21 )
```

File: src/hostname/hostnamed.c [hunk 5]
```diff
@@ -1128,7 +1158,7 @@ static int method_get_hardware_serial(sd_bus_message *m, void *userdata, sd_bus_
 
 static int method_describe(sd_bus_message *m, void *userdata, sd_bus_error *error) {
         _cleanup_free_ char *hn = NULL, *dhn = NULL, *in = NULL, *text = NULL,
-                *chassis = NULL, *vendor = NULL, *model = NULL, *serial = NULL;
+                *chassis = NULL, *vendor = NULL, *model = NULL, *serial = NULL, *firmwareversion = NULL;
```
Comment (bluca) [inline src/hostname/hostnamed.c]: style nit: this should be `firmware_version` to match other functions

File: src/hostname/hostnamed.c [hunk 6]
```diff
@@ -235,6 +233,20 @@ static int get_hardware_serial(char **ret) {
         return !isempty(s);
 }
 
+static int get_hardware_serial(char **ret) {
+         int r;
+
+         r = get_hardware_firmware_data("product_serial", ret);
+         if (r < 0)
```
Comment (yuwata) [inline src/hostname/hostnamed.c]: This must be `r <= 0`.

File: src/hostname/hostnamed.c [hunk 7]
```diff
@@ -601,6 +613,24 @@ static int property_get_hardware_model(
...
 
+static int property_get_firmware_version(
+                sd_bus *bus,
+                const char *path,
+                const char *interface,
+                const char *property,
+                sd_bus_message *reply,
+                void *userdata,
+                sd_bus_error *error) {
+        _cleanup_free_ char *firmware_version = NULL;
```
Comment (yuwata) [inline src/hostname/hostnamed.c:625]: Please insert a blank line before the variable declaration if function arguments are wrapped.

File: src/hostname/hostnamed.c [hunk 8]
```diff
@@ -208,21 +208,19 @@ static int get_hardware_model(char **ret) {
         return get_dmi_data("ID_MODEL_FROM_DATABASE", "ID_MODEL", ret);
 }
 
-static int get_hardware_serial(char **ret) {
+static int get_hardware_firmware_data(const char *dmi_data, char **ret) {
```
Comment (yuwata) [inline src/hostname/hostnamed.c]: Not a matter, but I prefer `dmi_data` -> `attribute`, `attr`, or `sysattr`.

File: src/hostname/hostnamectl.c [hunk 9]
```diff
@@ -299,6 +308,7 @@ static int show_all_names(sd_bus *bus) {
                 { "HomeURL",                   "s", NULL, offsetof(StatusInfo, home_url)        },
                 { "HardwareVendor",            "s", NULL, offsetof(StatusInfo, hardware_vendor) },
                 { "HardwareModel",             "s", NULL, offsetof(StatusInfo, hardware_model)  },
+                { "FirmwareVersion",           "s", NULL, offsetof(StatusInfo, firmware_version)},
```
Comment (yuwata) [inline src/hostname/hostnamectl.c]: Missing white space at the end.

**Comment by bluca** [comment]
Please rebase and squash the fixup commits, and force push

**Review by yuwata — CHANGES_REQUESTED** [review]
Looks mostly good. But one critical comment. Others are not so important.

**Comment by bluca** [comment]
The commit message should be prefixed with the component, eg: `hostnamed: display firmware version`

**Comment by cerebro1** [comment]
> > Output on my system : `sonalis@cere:~/outreachyspring2022/systemd [hostname_outreachy] $ cat build/test/sys/class/dmi/id/bios_version 7IET31WW (1.12 )` Output of hostnamectl : `sonalis@cere:~/outreachyspring2022/systemd [hostname_outreachy] $ ./build/hostnamectl Static hostname: cere Icon name: computer-laptop Chassis: laptop 💻 Machine ID: e7250c9873e54839aa321e08d41dbf30 Boot ID: 5029ec53bf7b45caa8a2ff2490887a70 Operating System: Ubuntu 18.04.3 LTS Kernel: Linux 4.15.0-175-generic Architecture: x86-64` Still value of Firmware Version not displayed by hostnamectl. Property return empty.
> 
> Seems to work fine here on my machine (in nspawn):
> 
> ```
> root@image:~# cat /sys/class/dmi/id/bios_version
> N27ET35W (1.21 )
> root@image:~# busctl get-property org.freedesktop.hostname1 /org/freedesktop/hostname1 org.freedesktop.hostname1 FirmwareVersion
> s "N27ET35W (1.21 )"
> root@image:~# hostnamectl
>    Static hostname: n/a                             
> Transient hostname: image
>          Icon name: computer
>         Machine ID: xxxx
>            Boot ID: 96aaafdfc98f49f0ba3ade1957bfd2aa
>     Virtualization: systemd-nspawn
>   Operating System: Debian GNU/Linux bookworm/sid   
>             Kernel: Linux 5.10.0-13-amd64
>       Architecture: x86-64
>   Firmware Version: N27ET35W (1.21 )
> ```

Thanks for sharing that. I was trying to running hostnamectl using build file. Is this an output from test image you built ?

**Comment by cerebro1** [comment]
> The commit message should be prefixed with the component, eg: `hostnamed: display firmware version`

yes. Can i like squash my commits and update message ?

**Comment by cerebro1** [comment]
> Please rebase and squash the fixup commits, and force push

ok

**Comment by bluca** [comment]
> > The commit message should be prefixed with the component, eg: `hostnamed: display firmware version`
> 
> yes. Can i like squash my commits and update message ?

yes, there are many ways to do that, rebase + squash first and then `git commit --amend` is the simplest

**Comment by bluca** [comment]
> > > Output on my system : `sonalis@cere:~/outreachyspring2022/systemd [hostname_outreachy] $ cat build/test/sys/class/dmi/id/bios_version 7IET31WW (1.12 )` Output of hostnamectl : `sonalis@cere:~/outreachyspring2022/systemd [hostname_outreachy] $ ./build/hostnamectl Static hostname: cere Icon name: computer-laptop Chassis: laptop 💻 Machine ID: e7250c9873e54839aa321e08d41dbf30 Boot ID: 5029ec53bf7b45caa8a2ff2490887a70 Operating System: Ubuntu 18.04.3 LTS Kernel: Linux 4.15.0-175-generic Architecture: x86-64` Still value of Firmware Version not displayed by hostnamectl. Property return empty.
> > 
> > 
> > Seems to work fine here on my machine (in nspawn):
> > ```
> > root@image:~# cat /sys/class/dmi/id/bios_version
> > N27ET35W (1.21 )
> > root@image:~# busctl get-property org.freedesktop.hostname1 /org/freedesktop/hostname1 org.freedesktop.hostname1 FirmwareVersion
> > s "N27ET35W (1.21 )"
> > root@image:~# hostnamectl
> >    Static hostname: n/a                             
> > Transient hostname: image
> >          Icon name: computer
> >         Machine ID: xxxx
> >            Boot ID: 96aaafdfc98f49f0ba3ade1957bfd2aa
> >     Virtualization: systemd-nspawn
> >   Operating System: Debian GNU/Linux bookworm/sid   
> >             Kernel: Linux 5.10.0-13-amd64
> >       Architecture: x86-64
> >   Firmware Version: N27ET35W (1.21 )
> > ```
> 
> Thanks for sharing that. I was trying to running hostnamectl using build file. Is this an output from test image you built ?

Yes, hostnamed is updated so that needs to be updated too - it is possible to run it on the host, but chances of breaking something are non-zero, so safer to run in an image. See this document when it talks about mkosi: https://github.com/systemd/systemd/blob/main/docs/HACKING.md

File: src/hostname/hostnamed.c (hunk 8 above)
Comment (cerebro1) [reply src/hostname/hostnamed.c]: preferred `sysattr`

File: src/hostname/hostnamectl.c (hunk 9 above)
Comment (cerebro1) [reply src/hostname/hostnamectl.c]: added

File: src/hostname/hostnamectl.c [hunk 10]
```diff
@@ -299,6 +308,7 @@ static int show_all_names(sd_bus *bus) {
                 { "HomeURL",                   "s", NULL, offsetof(StatusInfo, home_url)        },
                 { "HardwareVendor",            "s", NULL, offsetof(StatusInfo, hardware_vendor) },
                 { "HardwareModel",             "s", NULL, offsetof(StatusInfo, hardware_model)  },
+                { "FirmwareVersion",           "s", NULL, offsetof(StatusInfo, firmware_version) },
```
Comment (bluca) [inline src/hostname/hostnamectl.c:311]: please re-break the whole table (ie, re-align the closing brackets in the lines above)

**Comment by yuwata** [comment]
Please do not include merge commit in PR.

**Comment by bluca** [comment]
if you do `git rebase -i systemd/main` and then in the editor that opens up, change every `pick` to `fixup` apart from the first line, it should rebase and squash all commits into a single one, rebased on latest main

File: src/hostname/hostnamectl.c (hunk 10 above)
Comment (cerebro1) [reply src/hostname/hostnamectl.c:311]: done

**Comment by cerebro1** [comment]
> if you do `git rebase -i systemd/main` and then in the editor that opens up, change every `pick` to `fixup` apart from the first line, it should rebase and squash all commits into a single one, rebased on latest main

yes. was working on that. Actually there was a new push from someone so my commit was no more latest.
In rebase, `reword` worked to update commit message.

Thanks a lot for help and guidance in this whole task. @bluca @yuwata @poettering

**Comment by yuwata** [comment]
Hm?? Then, now this include irrelevant commits...
Please try something like the following
```
$ git fetch upstream main  <-- please adjust the repository name
$ git rebase upstream/main <-- here also
$ git push origin hostname_outreachy --force
```

File: src/hostname/hostnamectl.c (hunk 10 above)
Comment (yuwata) [reply src/hostname/hostnamectl.c:311]: Not aligned yet...

**Comment by yuwata** [comment]
Now it is better, but one more commit at the end!

**Comment by cerebro1** [comment]
> Now it is better, but one more commit at the end!
Well this will definately build my git knowledge. Sorry for the mess
Done this : 
git push --set-upstream origin hostname_outreachy

**Comment by yuwata** [comment]
Now, you can simply drop the last commit with `git reset --hard HEAD^`.

**Comment by bluca** [comment]
> Now, you can simply drop the last commit with `git reset --hard HEAD^`.

followed by `git push -f`

**Comment by bluca** [comment]
This should make it easier for next time, configure git globally to rebase when pulling, instead of merging (it really should be the default...):

```
git config --global pull.rebase true
```

**Comment by cerebro1** [comment]
> This should make it easier for next time, configure git globally to rebase when pulling, instead of merging (it really should be the default...):
> 
> ```
> git config --global pull.rebase true
> ```

Thanks a lot @bluca @yuwata . I can see i am out of touch with git. Will learn and try not repeating this. Yes updating config now.

**Comment by yuwata** [comment]
Then, remaining issue is only the table alignment. Almost perfect now.

**Comment by bluca** [comment]
> Then, remaining issue is only the table alignment. Almost perfect now.

it looks fine now?

**Comment by bluca** [comment]
> > Then, remaining issue is only the table alignment. Almost perfect now.
> 
> it looks fine now?

oh it doesn't anymore, I guess the fix was in a commit that was dropped

**Comment by cerebro1** [comment]
> > > Then, remaining issue is only the table alignment. Almost perfect now.
> > 
> > 
> > it looks fine now?
> 
> oh it doesn't anymore, I guess the fix was in a commit that was dropped

should i touch it or leave it because i was planning on sd-bus macros also of hostname.

**Comment by bluca** [comment]
> > > > Then, remaining issue is only the table alignment. Almost perfect now.
> > > 
> > > 
> > > it looks fine now?
> > 
> > 
> > oh it doesn't anymore, I guess the fix was in a commit that was dropped
> 
> should i touch it or leave it because i was planning on updating sd-bus macros also of hostname to the latest.

Please fix it - you can simply do the change, then update the commit with `git commit -a --amend --no-edit` and then force push

**Comment by cerebro1** [comment]
> > > > > Then, remaining issue is only the table alignment. Almost perfect now.
> > > > 
> > > > 
> > > > it looks fine now?
> > > 
> > > 
> > > oh it doesn't anymore, I guess the fix was in a commit that was dropped
> > 
> > 
> > should i touch it or leave it because i was planning on updating sd-bus macros also of hostname to the latest.
> 
> Please fix it - you can simply do the change, then update the commit with `git commit -a --amend --no-edit` and then force push

yes, kinda looks bad to leave. Already done. Thanks!

**Review by yuwata — APPROVED** [review]
Perfect!

**Review by bluca — APPROVED** [review]

**Comment by cerebro1** [comment]
hostnamectl is failing for a specific test. Debugging : 
```
======================================================================
ERROR: test_transient_hostname (__main__.DnsmasqClientTest)
networkd sets transient hostname from DHCP
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/autopkgtest.1QsS0h/build.A0G/systemd/test/networkd-test.py", line 787, in test_transient_hostname
    out = subprocess.check_output(['hostnamectl'])
  File "/usr/lib/python3.8/subprocess.py", line 415, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 516, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['hostnamectl']' returned non-zero exit status 1.

----------------------------------------------------------------------
Ran 35 tests in 146.245s

FAILED (errors=1, skipped=2)
autopkgtest [22:26:28]: test networkd-test.py: -----------------------]
autopkgtest [22:26:29]: test networkd-test.py:  - - - - - - - - - - results - - - - - - - - - -
networkd-test.py     FAIL non-zero exit status 1
```

**Comment by cerebro1** [comment]
> hostnamectl is failing for a specific test. Debugging :
> 
> ```
> ======================================================================
> ERROR: test_transient_hostname (__main__.DnsmasqClientTest)
> networkd sets transient hostname from DHCP
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/tmp/autopkgtest.1QsS0h/build.A0G/systemd/test/networkd-test.py", line 787, in test_transient_hostname
>     out = subprocess.check_output(['hostnamectl'])
>   File "/usr/lib/python3.8/subprocess.py", line 415, in check_output
>     return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
>   File "/usr/lib/python3.8/subprocess.py", line 516, in run
>     raise CalledProcessError(retcode, process.args,
> subprocess.CalledProcessError: Command '['hostnamectl']' returned non-zero exit status 1.
> 
> ----------------------------------------------------------------------
> Ran 35 tests in 146.245s
> 
> FAILED (errors=1, skipped=2)
> autopkgtest [22:26:28]: test networkd-test.py: -----------------------]
> autopkgtest [22:26:29]: test networkd-test.py:  - - - - - - - - - - results - - - - - - - - - -
> networkd-test.py     FAIL non-zero exit status 1
> ```

We already have this https://github.com/systemd/systemd/blob/main/src/network/systemd-networkd.rules
So why is the test failing ?

File: src/hostname/hostnamed.c [hunk 11]
```diff
@@ -601,6 +613,25 @@ static int property_get_hardware_model(
...
+                sd_bus_message *reply,
+                void *userdata,
+                sd_bus_error *error) {
+
+        _cleanup_free_ char *firmware_version = NULL;
+        int r;
+
+        r = get_firmware_version(&firmware_version);
+        if (r < 0)
+                return r;
```
Comment (yuwata) [inline src/hostname/hostnamed.c]: Ah, try to voidify the call. Some archs may not have the dmi devices.

Comment (cerebro1) [reply src/hostname/hostnamed.c]: ok. i thought it's related to this https://github.com/systemd/systemd/issues/4753. Will try to voidify it.

Comment (cerebro1) [reply src/hostname/hostnamed.c]: Will squash commits if ci passes

Comment (cerebro1) [reply src/hostname/hostnamed.c]: done

**Comment by poettering** [comment]
Looks excellent! lgtm

**Comment by cerebro1** [comment]
> Hm?? Then, now this includes irrelevant commits... Please try something like the following
> 
> ```
> $ git fetch upstream main  <-- please adjust the repository name 'upstream'
> $ git rebase upstream/main <-- here also
> $ git push origin hostname_outreachy --force
> ```

This saved me today :)

**Pull request merged**